### PR TITLE
keith/OFFAPPS-1087

### PR DESCRIPTION
**Pull request by keitheous**
OFFAPPS-1087 - Don't open menu when all options selected

## Description
The multi-select dropdown shows a list of options for groups in Zendesk Support.
When all groups are selected, the dropdown box stills appear but shows an empty list.

## Tasks
- [ ] Rebase Master to fast forward branch
- [x] Npm run test & lint.
- [x] Fix `src/javascript/datalist.js line 120` that isn't working properly.
- [x] Prevent `$menu.setAttribute('aria-hidden', false)` and `this.trigger('open_menu')` when all options are already selected.

## Implementation notes
- `src/javascript/datalist.js line 120` that isn't working properly as this is never called. Allocate the right event containing the class. 

- Created a check called `allOptionsSelected` to check for options attribute `isSelected`. If all is Selected. Prevent `trigger('open_menu')` and `$menu` from showing up.

- Kindly refers to screenshots

## References
[JIRA ticket](https://zendesk.atlassian.net/browse/OFFAPPS-1087)

## Screenshots of Implementation Result
<details>
  <summary>Indicator</summary>
 <br><em>When the arrow points up, the tray is open to display unselected user groups options.</em><br>
11 copy

 <br><em>When the arrow points down, the tray is close to hide options.</em><br>
111 copy
</details>

<details>
  <summary>Admin Page</summary>
 <br><em>Step 1: Observe all user groups in options tray</em><br>

 <br><em>Step 2: Select all user groups</em><br>

 <br><em>Step 3: Do not display empty options anymore when all selected</em><br>

 <br><em>Step 4: Browse options after removing one user group</em><br>

</details>

<details>
  <summary>Broadcast Page</summary>
 <br><em>Step 1: Click into broadcast page</em><br>

 <br><em>Step 2: Select one of the user groups</em><br>

 <br><em>Step 3: Do not display empty options anymore when all selected </em><br>

 <br><em>Step 4:</em><br>

</details>

## CCs
@zendesk/apps-migration 

## Risks
* [low] Drop down list for groups in both broadcast and admin page. 100% unlikely the tray doesn't open up

### CODE REVIEW

File: src/javascript/lib/datalist.js [hunk 1]
```diff
@@ -190,12 +190,20 @@ export default class Datalist {
...
 
-    this.trigger('open_menu')
+    if (!this.allOptionsSelected()) {
+      $menu.setAttribute('aria-hidden', false)
+      this.trigger('open_menu')
+    }
+  }
+
+  allOptionsSelected () {
+    return this.options.reduce((accumulation, option) => {
```
Comment (yu-tian113) [inline src/javascript/lib/datalist.js]: I feel `every` is more appropriate than `reduce` here? What you think?
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/every

File: src/javascript/lib/datalist.js [hunk 2]
```diff
@@ -117,7 +117,7 @@ export default class Datalist {
     })
 
     // Skip if clicked to remove a tag
-    if (event.eventTarget.classList.contains('c-tag__remove')) return
+    if (event.target.classList.contains('c-tag__remove')) return
```
Comment (yu-tian113) [inline src/javascript/lib/datalist.js]: I believe @ocke added this `eventTarget` property to handle event bubbling.
Is this change intentional?

File: src/javascript/lib/datalist.js (hunk 1 above)
Comment (keitheous) [reply src/javascript/lib/datalist.js]: screen shot 2018-11-07 at 11 55 09 am

That's pretty COOL! Thanks for the recommendation!

File: src/javascript/lib/datalist.js (hunk 2 above)
Comment (keitheous) [reply src/javascript/lib/datalist.js]: If I can, I would like to book you in (@yu-tian113) and go through this with you.

Comment (yu-tian113) [reply src/javascript/lib/datalist.js]: 👍

Comment (vaevictis) [reply src/javascript/lib/datalist.js]: Curious to know how this went 👀. I don't understand why there would be need for a custom target to be added.

Comment (keitheous) [reply src/javascript/lib/datalist.js]: Tony and I sat down and discussed this in person.

We were trying to step into Olaf's shoes to figure if this was intended to be either event.target or event.eventTarget (helper method). Why and which made more sense.

We concluded that eventTarget, in this case, will always be the containing element (.selector) due to the event delegation of const events. event.target made more sense in this case.

I can sit down and go this with you if you'd like @vaevictis .

Comment (vaevictis) [reply src/javascript/lib/datalist.js]: I think I'm fine with your explanation @keitheous. Thanks!

File: src/javascript/lib/datalist.js [hunk 3]
```diff
@@ -201,9 +201,7 @@ export default class Datalist {
   }
 
   allOptionsSelected () {
-    return this.options.reduce((accumulation, option) => {
-      return accumulation && option.isSelected
-    }, true)
+    return this.options.every((option) => option.isSelected === true)
```
Comment (vaevictis) [inline src/javascript/lib/datalist.js]: ❤️

Comment (keitheous) [reply src/javascript/lib/datalist.js]: ❤️
